pkg/evaluator/evaluator.go

```diff
@@ -213,6 +215,94 @@ func (e *Evaluator) evalWhile(scope *scope, w *parser.While) Value {
 	return val
 }
 
+func (e *Evaluator) evalFor(scope *scope, f *parser.For) Value {
+	scope = newInnerScope(scope)
+	r, err := e.newRange(scope, f)
+	if err != nil {
+		return err
+	}
+	for r.next() {
+		val := e.Eval(scope, f.Block)
+		if isError(val) || isBreak(val) || isReturn(val) {
+			return val
+		}
+	}
+	return nil
+}
+
+func (e *Evaluator) newRange(scope *scope, f *parser.For) (ranger, Value) {
+	if r, ok := f.Range.(*parser.StepRange); ok {
+		return e.newStepRange(scope, r, f.LoopVar)
+	}
+	rangeVal := e.Eval(scope, f.Range)
+	if isError(rangeVal) {
+		return nil, rangeVal
+	}
+
+	switch v := rangeVal.(type) {
+	case *Array:
+		loopVar := zero(f.LoopVar.Type())
+		scope.set(f.LoopVar.Name, loopVar)
+		return &arrayRange{loopVar: loopVar, array: v, cur: 0}, nil
+	case *String:
+		loopVar := &String{}
+		scope.set(f.LoopVar.Name, loopVar)
+		return &stringRange{loopVar: loopVar, str: v, cur: 0}, nil
+	case *Map:
+		loopVar := &String{}
+		scope.set(f.LoopVar.Name, loopVar)
+		order := make([]string, len(*v.Order))
+		copy(order, *v.Order)
+		m := &mapRange{loopVar: loopVar, mapVal: v, cur: 0, order: order}
+		return m, nil
+	}
+	return nil, newError("cannot create range for " + f.Range.String())
+}
+
+func (e *Evaluator) newStepRange(scope *scope, r *parser.StepRange, loopVar *parser.Var) (ranger, Value) {
+	start, err := e.numValWithDefault(scope, r.Start, 0.0)
+	if err != nil {
+		return nil, err
+	}
+	stop, err := e.numVal(scope, r.Stop)
+	if err != nil {
+		return nil, err
+	}
+	step, err := e.numValWithDefault(scope, r.Step, 1.0)
```

Comment:
hmmm, the second return value is of type `Value` not `error`, so calling it `err` is a little misleading. I know it is the error case, but still... Perhaps call it errValue or something?

Reply:
fair, done.